### PR TITLE
Login changes

### PR DESCRIPTION
**Pull request by alesstimec**
Login changes

Since JAAS is moving away from candid to OIDC, juju clients need to be able to log in to JIMM using Admin facade version 4 that implements new methods to allow log in using either an access token or client credentials. Additional methods are used to initiate the device oauth2 flow to obtain a valid access token.

1. Introduces a LoginProvider that logs in with the juju controller once the websock connection is established:
  - implemented a login provider for the current login that uses username, password or macaroons
  - implemented a login provider that uses provided client credentials
  - implemented a login provider that uses an access token
2. Adds an additional field to AccountDetails (an access token). Changing the way we parse accounts.yaml to something other than AccountDetails (to be more in-line with credentials.yaml) would be a massive undertaking because of the many places that use the AccountDetails - one possibly done in a followup.

## Checklist

- [ ] Code style: imports ordered, good names, simple structure, etc
- [ ] Comments saying why design decisions were made
- [x  ] Go unit tests, with comments saying what you're testing
- [ ] [Integration tests](https://github.com/juju/juju/tree/main/tests), with comments saying what you're testing
- [ ] [doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages

## Links

**Launchpad bug:** https://bugs.launchpad.net/juju/+bug/

**Jira card:** JUJU-

### CODE REVIEW

File: jujuclient/file.go [hunk 1]
```diff
@@ -37,7 +37,7 @@ var (
 
 // NewFileClientStore returns a new filesystem-based client store
 // that manages files in $XDG_DATA_HOME/juju.
-func NewFileClientStore() ClientStore {
+func NewFileClientStore() *store {
```
Comment (wallyworld) [inline jujuclient/file.go:40]: What's the reason for this change?

File: jujuclient/interface.go [hunk 2]
```diff
@@ -101,6 +101,8 @@ type AccountDetails struct {
 	// They are only set when using the MemStore implementation,
 	// and are used by embedded commands. The are not written to disk.
 	Macaroons []macaroon.Slice `yaml:"-"`
+
+	AccessToken string `yaml:"access-token,omitempty"`
```
Comment (wallyworld) [inline jujuclient/interface.go:105]: Needs a comment

Comment (wallyworld) [reply jujuclient/interface.go:105]: I thought we discussed adding a "login-type" or "login-provider" attribute and we'd use a factory type pattern to instantiate the login provider to use with the state (renamed to conn in juju 4.0 BTW)

File: api/apiclient.go [hunk 3]
```diff
@@ -329,7 +237,19 @@ func PerferredHost(info *Info) string {
...
+		// which logs in using username/password or macaroons.
+		if st.loginProvider == nil {
+			result <- errors.Errorf("login provider not specified")
+		}
+
+		loginResult, err := st.loginProvider.Login(ctx, st)
+		if err != nil {
+			result <- err
+			return
+		}
```
Comment (wallyworld) [inline api/apiclient.go:250]: I think we could improve this - we're using `st` to get the provider and then also as a parameter to Login. At the time we parse the account details, can we set up st so that we just need to then call `st.Login(info)` or something. In fact, I can't see where the info passed into loginWithContext is used.

So spitballing, something like (not necessarily this exactly)

```
func loginWithContext(ctx context.Context, st *conn, info *Info) error {
	result := make(chan error, 1)
	go func() {
		result <- st.Login(ctx, info)
	}()
	select {
	case err := <-result:
		return errors.Trace(err)
	case <-ctx.Done():
		return errors.Annotatef(ctx.Err(), "cannot log in")
	}
}
```
The above keeps the login provider abstract hidden from the login flow. And we aren't cherry picking attributes off info to pass in (which are provider dependent).

File: jujuclient/interface.go (hunk 2 above)
Comment (alesstimec) [reply jujuclient/interface.go:105]: we did discuss it.. but, as my comment in the PR says, that is a much much larger change due to the many places where AccountDetails are used. And it doesn't give us much, since we did say that we want to use the same accounts.yaml file and be backwards compatible. Well this simple addition is backwards compatible, so a 2.9 client is still able to read the accoutns.yaml file..  And the change to use a factory type pattern should possibly be done in a followup to keep PRs small and focused

File: api/apiclient.go (hunk 3 above)
Comment (alesstimec) [reply api/apiclient.go]: @wallyworld ptal - changed: st does not need to hold the loginProvider as it only holds the state once login completes.. also Info is not needed in the loginWithContext call..

File: jujuclient/file.go (hunk 1 above)
Comment (alesstimec) [reply jujuclient/file.go]: reverted

File: api/state.go [hunk 4]
```diff
@@ -4,187 +4,135 @@
...
+
+	// proxier is the proxier used for this connection when not nil. If's expected
+	// the proxy has already been started when placing in this var. This struct
+	// will take the responsibility of closing the proxy.
+	proxier jujuproxy.Proxier
+}
+
+// Login implements the Login method of the Connection interface providing authentication
+// using basic auth or macaroons.
+func (st *state) Login(name names.Tag, password, nonce string, ms []macaroon.Slice) error {
```
Comment (tlm) [inline api/state.go:124]: Should we not be removing this function in favour of a login provider?

**Comment by tlm** [comment]
FYI Brain Dump,

This is a placeholder for anyone finding this PR in the future and wondering why some decisions were made here. Today the JAAS folks are kindly doing some work for Juju to start enabling OAuth device flows. Due to legacy reasons around how the clients Login flow works this has been a tricky change to implement. This PR will not address all of the design considerations that we would normally expect from a Juju PR as it requires work that we currently don't have the bandwidth for. @wallyworld and @tlm have decided to sail this through in a slightly bad design state knowing that we have to come back and do some bigger surgery. We can't ask the JAAS guys to deliver these design changes for us at the moment.


## Changes to accounts.yaml
We have thrown back and forward the idea of introducing an accounts2.yaml file into the Juju client to better describe the different authn flows that we now have. This is a change we still plan on going ahead with but not in this PR. The main reason is there are far too many places that's touch accounts.yaml for this PR to fully make the switch.

We would plan on introducing a `type`field to each of the accounts in the yaml field that can be used as a key into a login provider factory to construct and manage the specific account types contained within this file. We would also need a way to upgrade legacy clients to the newer accounts file and possibly even maintain backwards compatibility for the older file if users wanted to downgrade their `juju` client version.

For this PR we have settled on a flat structure of just introducing the type field and a few more fields to describe the oauth specific keys. Best of both worlds for the moment.


## Removing Basic Auth Fields From Info
In `api/apiclient.go` these lines exists within `func Open(info *Info, opts DialOpts) (Connection, error)`
```go
       loginProvider := opts.LoginProvider
	if loginProvider == nil {
		loginProvider = NewUserpassLoginProvider(info.Tag, info.Password, info.Nonce, info.Macaroons, bakeryClient, CookieURLFromHost(host))
	}
```

We want to move away from this and remove any authn credentials from the info struct and error if `opts.LoginProvider == nil`.  The login provider should be created further up the stack and be a closure around the specific secret bits that pertain to that login method.

## Remove go func from `loginWithContext`
There exists a goroutine in `loginWithContext` that is very complicated and doesn't seem to be adding and value when the caller of the goroutine is blocking on the result anyway.

## Tests
We are missing a lot of tests in this area that need to be dealt with quickly. @alesstimec If you are able to provide any tests for the JAAS parts that would be greatly appreciated.

# Conclusion
Overall the direction we are heading here is better then what we had but we need to double down a little more and fully embrace the encapsulated login providers from the rest of the api dealing code. We also need to pay close attention to the error messages returned to the user. At the moment I think we are possible just wrapping errors up the stack not giving much thought as to what the end user is going to see.

This is a brain dump so that @tlm or @wallyworld can come back soon and polish this up some more.

File: api/apiclient.go [hunk 5]
```diff
@@ -329,7 +237,19 @@ func PerferredHost(info *Info) string {
 func loginWithContext(ctx context.Context, st *state, info *Info) error {
 	result := make(chan error, 1)
 	go func() {
-		result <- st.Login(info.Tag, info.Password, info.Nonce, info.Macaroons)
+		// if loginProvider is not specified we use the default login provider
+		// which logs in using username/password or macaroons.
+		if st.loginProvider == nil {
+			result <- errors.Errorf("login provider not specified")
```
[the remark this answers is not in the document]
Comment (alesstimec) [reply api/apiclient.go]: done

File: api/connector/simpleconnector.go [hunk 6]
```diff
@@ -27,6 +27,11 @@ type SimpleConfig struct {
 	Username  string
 	Password  string
 	Macaroons []macaroon.Slice
+
+	// ClientID and ClientSecret represent the client credentials
+	// created with the IdP.
```
[the remark this answers is not in the document]
Comment (alesstimec) [reply api/connector/simpleconnector.go]: added docs

File: api/login.go [hunk 7]
```diff
@@ -0,0 +1,395 @@
...
+	"github.com/juju/juju/rpc"
+	"github.com/juju/juju/rpc/params"
+	jujuversion "github.com/juju/juju/version"
+)
+
+var (
+	loginDeviceAPICall = func(caller base.APICaller, request interface{}, response interface{}) error {
+		return caller.APICall("Admin", 4, "", "LoginDevice", request, response)
+	}
+	getDeviceSessionTokenAPICall = func(caller base.APICaller, request interface{}, response interface{}) error {
```
[the remark this answers is not in the document]
Comment (alesstimec) [reply api/login.go:32]: we are obtaining a session token from JIMM - JIMM never gets access to user's credentials, only user's id token, once the user visits the verification url, enters the user code and logs in.

File: api/state.go (hunk 4 above)
Comment (alesstimec) [reply api/state.go:124]: removed, but added a convenience function to export_test.go

Comment (alesstimec) [reply api/state.go:124]: Actually.. let's leave this in as a convenient method to log in using the user-pass login provider.. there are quite a few tests in apiserver/admin_test.go that test various modes of log in using the username-password/macaroons/nonce.. most of them open a websocket connection then log in after the fact..

File: api/login.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (tlm) [reply api/login.go:32]: Thanks

File: api/connector/simpleconnector.go [hunk 8]
```diff
@@ -53,10 +58,20 @@ func NewSimple(opts SimpleConfig, dialOptions ...api.DialOption) (*SimpleConnect
 	if err := info.Validate(); err != nil {
 		return nil, err
 	}
+
+	dialOpts := api.DefaultDialOpts()
+	if opts.ClientID != "" && opts.ClientSecret != "" {
```
[the remark this answers is not in the document]
Comment (tlm) [reply api/connector/simpleconnector.go:64]: But we are going to be storing other information in accounts.yaml like an access token once the user has logged in with the auth server?

**Review by tlm — APPROVED** [review]
Thank you, just wanting to see if we should be introducing the type field now?

File: api/connector/simpleconnector.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (alesstimec) [reply api/connector/simpleconnector.go:64]: no, the terraform provider will send client id and secret every time - no session tokens.. client credentials are meant for machine2machine communication

File: api/apiclient.go [hunk 9]
```diff
@@ -326,10 +238,21 @@ func PerferredHost(info *Info) string {
...
 // that this becomes unnecessary.
-func loginWithContext(ctx context.Context, st *state, info *Info) error {
+func loginWithContext(ctx context.Context, loginProvider LoginProvider, st *state) error {
 	result := make(chan error, 1)
 	go func() {
-		result <- st.Login(info.Tag, info.Password, info.Nonce, info.Macaroons)
+		if loginProvider == nil {
+			result <- errors.New("login provider not specified")
+			return
+		}
```
Comment (wallyworld) [inline api/apiclient.go:247]: This check doesn't need to be in the go routine. It should just be done as part of loginWithContext args checking. 
Would prefer loginProvider to be the 3rd arg.

File: api/login.go [hunk 10]
```diff
@@ -0,0 +1,440 @@
...
+	updateAccountDetailsFunc func(string) error,
+) *sessionTokenLoginProvider {
+	return &sessionTokenLoginProvider{
+		sessionToken:             token,
+		printOutputFunc:          printOutputFunc,
+		updateAccountDetailsFunc: updateAccountDetailsFunc,
+	}
+}
+
+type sessionTokenLoginProvider struct {
```
Comment (wallyworld) [inline api/login.go:57]: Can we move these login provider implementations to a different file `loginproviders.go`, same with tests.
Or better a separate go file per provider to keep things nice and neat.

File: api/state.go (hunk 4 above)
Comment (wallyworld) [reply api/state.go:124]: Longer term we want to clean this up and have the login providers do it. But to make progress, can we add a TODO to refactor.

File: api/login.go [hunk 11]
```diff
@@ -0,0 +1,440 @@
...
+// of the first will be returned.
+// This login provider should only be used when connecting to a controller
+// for the first time when we still don't know which login method.
+func NewAnyOfLoginProvider(providers ...LoginProvider) LoginProvider {
+	return &anyOfLoginProviders{
+		providers: providers,
+	}
+}
+
+type anyOfLoginProviders struct {
```
Comment (wallyworld) [inline api/login.go:425]: Can we call this `orderedLoginProviders` or `tryInOrderLoginProviders` or `prioritisedLoginProviders` or something to make it clear order matters. `Any` implies try them in some undetermined order and whatever gets there is used.

Comment (wallyworld) [reply api/login.go:425]: Ideally we'd probe an unauthenticated login endpoint to get an ordered list of supported login types. Have we thought about that approach? And if we get back a "not implemented" error we fall back to UserPass.

**Review by wallyworld — APPROVED** [review]
We can make progress. But I think we need to follow up with a refactoring to clean up login provider semantics and IMO we need to look at probing for supported login types and not just checking stuff at the wall and see what sticks. The server should be the one to inform the client what's supported.

File: api/login.go (hunk 11 above)
Comment (alesstimec) [reply api/login.go:425]: we are doing something similar with this approach.. trying oauth2 first, if not implemented falling back to user-pass.. would a separate endpoint have any value? are we planning on supporting more login providers?

File: api/apiclient.go (hunk 9 above)
Comment (alesstimec) [reply api/apiclient.go]: done

File: api/login.go (hunk 10 above)
Comment (alesstimec) [reply api/login.go]: done

File: juju/api.go [hunk 12]
```diff
@@ -156,10 +156,11 @@ func NewAPIConnection(args NewAPIConnectionParams) (_ api.Connection, err error)
 				accountDetails.LastKnownAccess = st.ControllerAccess()
 			}
 		}
-		if ok && !user.IsLocal() && apiInfo.Tag == nil {
+		if ok && !user.IsLocal() && apiInfo.Tag == nil && accountDetails.Type != jujuclient.OAuth2DeviceFlowAccountDetailsType {
```
Comment (kian99) [inline juju/api.go:159]: On this line, accountDetails can be nil leading to a null pointer dereference. This is possible if the Juju CLI is running without a store.

Comment (alesstimec) [reply juju/api.go]: fixed

File: api/login.go (hunk 11 above)
Comment (wallyworld) [reply api/login.go]: We are doing it ordered as you say, but the name of the struct does not reflect that and implies to the casual observed that order does not matter.
As discussed elsewhere, allowing the client probe the auth server for what is supported is a cleaner approach that we should adopt. That way, the client does not waste time and resources trying oauth2 when the server does not even support it, eg older controllers. We can make progress by landing this but we should not go to release without the login probe being done.

**Comment by SimonRichardson** [comment]
@alesstimec as follow up can you move `NewSessionTokenLoginProvider` out of the API package? This shouldn't have been put in there and was a miss-step (also can you rename the file, we concatenate the names without underscores). You can put it in the `cmd/api` or `cmd/internal/api` package.

The API base package shouldn't have interactive commands and we're trying to ensure in the future that the API package is clean and appropriate for all use cases (3rd-parties, terraform, etc).
